File: lib/djvApp/Widgets/Viewport.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the DJV project.

#include <djvApp//Widgets/Viewport.h>

#include <djvApp/Models/ColorModel.h>
#include <djvApp/Models/FilesModel.h>
#include <djvApp/Models/SettingsModel.h>
#include <djvApp/Models/TimeUnitsModel.h>
#include <djvApp/Models/ViewportModel.h>
#include <djvApp/App.h>

#include <tlRender/Timeline/Util.h>

#include <ftk/UI/ColorSwatch.h>
#include <ftk/UI/GridLayout.h>
#include <ftk/UI/Label.h>
#include <ftk/UI/RowLayout.h>
#include <ftk/UI/Spacer.h>
#include <ftk/Core/Format.h>

#include <regex>

namespace djv
{
    namespace app
    {
        struct Viewport::Private
        {
            std::weak_ptr<App> app;
            bool hud = false;
            ftk::Path path;
            OTIO_NS::RationalTime currentTime = tl::invalidTime;
            double fps = 0.0;
            size_t droppedFrames = 0;
            size_t videoFramesSize = 0;
            ftk::ImageOptions imageOptions;
            tl::DisplayOptions displayOptions;
            tl::PlayerCacheInfo cacheInfo;
            MouseActionBinding pickBinding =
                MouseActionBinding(ftk::MouseButton::Left);
            MouseActionBinding frameShuttleBinding =
                MouseActionBinding(ftk::MouseButton::Left, ftk::KeyModifier::Shift);
            float frameShuttleScale = 1.F;
            std::shared_ptr<ftk::Observable<ftk::V2I> > pick;
            std::shared_ptr<ftk::Observable<ftk::V2I> > samplePos;
            std::shared_ptr<ftk::Observable<ftk::Color4F> > colorSample;

            std::shared_ptr<ftk::Label> fileNameLabel;
            std::shared_ptr<ftk::Label> timeLabel;
            std::shared_ptr<ftk::ColorSwatch> colorPickerSwatch;
            std::shared_ptr<ftk::Label> colorPickerLabel;
            std::shared_ptr<ftk::Label> cacheLabel;
            std::shared_ptr<ftk::GridLayout> hudLayout;

            std::shared_ptr<ftk::Observer<OTIO_NS::RationalTime> > currentTimeObserver;
            std::shared_ptr<ftk::ListObserver<tl::VideoFrame> > videoObserver;
            std::shared_ptr<ftk::Observer<tl::PlayerCacheInfo> > cacheObserver;
            std::shared_ptr<ftk::Observer<double> > fpsObserver;
            std::shared_ptr<ftk::Observer<size_t> > droppedFramesObserver;
            std::shared_ptr<ftk::Observer<tl::CompareOptions> > compareOptionsObserver;
            std::shared_ptr<ftk::Observer<tl::OCIOOptions> > ocioOptionsObserver;
            std::shared_ptr<ftk::Observer<tl::LUTOptions> > lutOptionsObserver;
            std::shared_ptr<ftk::Observer<ftk::ImageOptions> > imageOptionsObserver;
            std::shared_ptr<ftk::Observer<tl::DisplayOptions> > displayOptionsObserver;
            std::shared_ptr<ftk::Observer<tl::BackgroundOptions> > bgOptionsObserver;
            std::shared_ptr<ftk::Observer<tl::ForegroundOptions> > fgOptionsObserver;
            std::shared_ptr<ftk::Observer<ftk::ImageType> > colorBufferObserver;
            std::shared_ptr<ftk::Observer<bool> > hudObserver;
            std::shared_ptr<ftk::Observer<tl::TimeUnits> > timeUnitsObserver;
            std::shared_ptr<ftk::Observer<MouseSettings> > mouseSettingsObserver;

            enum class MouseMode
            {
                None,
                Shuttle,
                Picker
            };
            struct MouseData
            {
                MouseMode mode = MouseMode::None;
                OTIO_NS::RationalTime shuttleStart = tl::invalidTime;
            };
            MouseData mouse;
        };

        void Viewport::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            tl::ui::Viewport::_init(context, parent);
            FTK_P();

            p.app = app;

            p.pick = ftk::Observable<ftk::V2I>::create();
            p.samplePos = ftk::Observable<ftk::V2I>::create();
            p.colorSample = ftk::Observable<ftk::Color4F>::create();

            p.fileNameLabel = ftk::Label::create(context);
            p.fileNameLabel->setFontRole(ftk::FontRole::Mono);
            p.fileNameLabel->setMarginRole(ftk::SizeRole::MarginInside);
            p.fileNameLabel->setBackgroundRole(ftk::ColorRole::Overlay);

            p.timeLabel = ftk::Label::create(context);
            p.timeLabel->setFontRole(ftk::FontRole::Mono);
            p.timeLabel->setMarginRole(ftk::SizeRole::MarginInside);
            p.timeLabel->setBackgroundRole(ftk::ColorRole::Overlay);
            p.timeLabel->setHAlign(ftk::HAlign::Right);

            p.colorPickerSwatch = ftk::ColorSwatch::create(context);
            p.colorPickerSwatch->setSizeRole(ftk::SizeRole::MarginLarge);
            p.colorPickerLabel = ftk::Label::create(context);
            p.colorPickerLabel->setFontRole(ftk::FontRole::Mono);

            p.cacheLabel = ftk::Label::create(context);
            p.cacheLabel->setFontRole(ftk::FontRole::Mono);
            p.cacheLabel->setMarginRole(ftk::SizeRole::MarginInside);
            p.cacheLabel->setBackgroundRole(ftk::ColorRole::Overlay);
            p.cacheLabel->setHAlign(ftk::HAlign::Right);

            p.hudLayout = ftk::GridLayout::create(context, shared_from_this());
            p.hudLayout->setMarginRole(ftk::SizeRole::MarginSmall);
            p.hudLayout->setSpacingRole(ftk::SizeRole::SpacingSmall);
            p.fileNameLabel->setParent(p.hudLayout);
            p.hudLayout->setGridPos(p.fileNameLabel, 0, 0);
            p.timeLabel->setParent(p.hudLayout);
            p.hudLayout->setGridPos(p.timeLabel, 0, 2);

            auto spacer = ftk::Spacer::create(context, ftk::Orientation::Vertical, p.hudLayout);
            spacer->setStretch(ftk::Stretch::Expanding);
            p.hudLayout->setGridPos(spacer, 1, 1);

            auto hLayout = ftk::HorizontalLayout::create(context, p.hudLayout);
            p.hudLayout->setGridPos(hLayout, 2, 0);
            hLayout->setMarginRole(ftk::SizeRole::MarginInside);
            hLayout->setSpacingRole(ftk::SizeRole::SpacingSmall);
            hLayout->setBackgroundRole(ftk::ColorRole::Overlay);
            p.colorPickerSwatch->setParent(hLayout);
            p.colorPickerLabel->setParent(hLayout);
            p.cacheLabel->setParent(p.hudLayout);
            p.hudLayout->setGridPos(p.cacheLabel, 2, 2);

            p.fpsObserver = ftk::Observer<double>::create(
                observeFPS(),
                [this](double value)
                {
                    _p->fps = value;
                    _hudUpdate();
                });

            p.droppedFramesObserver = ftk::Observer<size_t>::create(
                observeDroppedFrames(),
                [this](size_t value)
                {
                    _p->droppedFrames = value;
                    _hudUpdate();
                });

            p.compareOptionsObserver = ftk::Observer<tl::CompareOptions>::create(
                app->getFilesModel()->observeCompareOptions(),
                [this](const tl::CompareOptions& value)
                {
                    setCompareOptions(value);
                });

            p.ocioOptionsObserver = ftk::Observer<tl::OCIOOptions>::create(
                app->getColorModel()->observeOCIOOptions(),
                [this](const tl::OCIOOptions& value)
                {
                   setOCIOOptions(value);
                });

            p.lutOptionsObserver = ftk::Observer<tl::LUTOptions>::create(
                app->getColorModel()->observeLUTOptions(),
                [this](const tl::LUTOptions& value)
                {
                   setLUTOptions(value);
                });

            p.imageOptionsObserver = ftk::Observer<ftk::ImageOptions>::create(
                app->getViewportModel()->observeImageOptions(),
                [this](const ftk::ImageOptions& value)
                {
                    _p->imageOptions = value;
                    _videoUpdate();
                });

            p.displayOptionsObserver = ftk::Observer<tl::DisplayOptions>::create(
                app->getViewportModel()->observeDisplayOptions(),
                [this](const tl::DisplayOptions& value)
                {
                    _p->displayOptions = value;
                    _videoUpdate();
                });

            p.bgOptionsObserver = ftk::Observer<tl::BackgroundOptions>::create(
                app->getViewportModel()->observeBackgroundOptions(),
                [this](const tl::BackgroundOptions& value)
                {
                    setBackgroundOptions(value);
                });

            p.fgOptionsObserver = ftk::Observer<tl::ForegroundOptions>::create(
                app->getViewportModel()->observeForegroundOptions(),
                [this](const tl::ForegroundOptions& value)
                {
                    setForegroundOptions(value);
                });

            p.colorBufferObserver = ftk::Observer<ftk::ImageType>::create(
                app->getViewportModel()->observeColorBuffer(),
                [this](ftk::ImageType value)
                {
                    setColorBuffer(value);
                    _hudUpdate();
                });

            p.hudObserver = ftk::Observer<bool>::create(
                app->getViewportModel()->observeHUD(),
                [this](bool value)
                {
                    _p->hud = value;
                    _hudUpdate();
                });

            p.timeUnitsObserver = ftk::Observer<tl::TimeUnits>::create(
                app->getTimeUnitsModel()->observeTimeUnits(),
                [this](tl::TimeUnits value)
                {
                    _hudUpdate();
                });

            p.mouseSettingsObserver = ftk::Observer<MouseSettings>::create(
                app->getSettingsModel()->observeMouse(),
                [this](const MouseSettings& value)
                {
                    FTK_P();
                    auto i = value.bindings.find(MouseAction::PanView);
                    setPanBinding(
                        i != value.bindings.end() ? i->second.button : ftk::MouseButton::None,
                        i != value.bindings.end() ? i->second.modifier : ftk::KeyModifier::None);
                    i = value.bindings.find(MouseAction::CompareWipe);
                    setWipeBinding(
                        i != value.bindings.end() ? i->second.button : ftk::MouseButton::None,
                        i != value.bindings.end() ? i->second.modifier : ftk::KeyModifier::None);
                    i = value.bindings.find(MouseAction::Pick);
                    p.pickBinding = i != value.bindings.end() ? i->second : MouseActionBinding();
                    i = value.bindings.find(MouseAction::FrameShuttle);
                    p.frameShuttleBinding = i != value.bindings.end() ? i->second : MouseActionBinding();
                    p.frameShuttleScale = value.frameShuttleScale;
                });
        }

        Viewport::Viewport() :
            _p(new Private)
        {}

        Viewport::~Viewport()
        {}

        std::shared_ptr<Viewport> Viewport::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            auto out = std::shared_ptr<Viewport>(new Viewport);
            out->_init(context, app, parent);
            return out;
        }

        std::shared_ptr<ftk::IObservable<ftk::V2I> > Viewport::observePick() const
        {
            return _p->pick;
        }

        std::shared_ptr<ftk::IObservable<ftk::V2I> > Viewport::observeSamplePos() const
        {
            return _p->samplePos;
        }

        std::shared_ptr<ftk::IObservable<ftk::Color4F> > Viewport::observeColorSample() const
        {
            return _p->colorSample;
        }

        void Viewport::setPlayer(const std::shared_ptr<tl::Player>& player)
        {
            tl::ui::Viewport::setPlayer(player);
            FTK_P();
            if (player)
            {
                p.path = player->getPath();

                p.currentTimeObserver = ftk::Observer<OTIO_NS::RationalTime>::create(
                    player->observeCurrentTime(),
                    [this](const OTIO_NS::RationalTime& value)
                    {
                        _p->currentTime = value;
                        _hudUpdate();
                    });

                p.videoObserver = ftk::ListObserver<tl::VideoFrame>::create(
                    player->observeCurrentVideo(),
                    [this](const std::vector<tl::VideoFrame>& value)
                    {
                        _p->videoFramesSize = value.size();
                        _videoUpdate();
                    });

                p.cacheObserver = ftk::Observer<tl::PlayerCacheInfo>::create(
                    player->observeCacheInfo(),
                    [this](const tl::PlayerCacheInfo& value)
                    {
                        _p->cacheInfo = value;
                        _hudUpdate();
                    });
            }
            else
            {
                p.path = ftk::Path();
                p.currentTime = tl::invalidTime;
                p.currentTimeObserver.reset();
                p.videoObserver.reset();
                p.cacheInfo = tl::PlayerCacheInfo();
                p.cacheObserver.reset();
                _hudUpdate();
            }
        }

        ftk::Size2I Viewport::getSizeHint() const
        {
            return _p->hudLayout->getSizeHint();
        }

        void Viewport::setGeometry(const ftk::Box2I& value)
        {
            tl::ui::Viewport::setGeometry(value);
            FTK_P();
            p.hudLayout->setGeometry(value);
        }

        void Viewport::mouseMoveEvent(ftk::MouseMoveEvent& event)
        {
            tl::ui::Viewport::mouseMoveEvent(event);
            FTK_P();
            switch (p.mouse.mode)
            {
            case Private::MouseMode::Shuttle:
                if (auto player = getPlayer())
                {
                    const OTIO_NS::RationalTime offset = OTIO_NS::RationalTime(
                        (event.pos.x - _getMousePressPos().x) * .05F * p.frameShuttleScale,
                        p.mouse.shuttleStart.rate()).round();
                    const OTIO_NS::TimeRange& timeRange = player->getTimeRange();
                    OTIO_NS::RationalTime t = p.mouse.shuttleStart + offset;
                    if (t < timeRange.start_time())
                    {
                        t = timeRange.end_time_exclusive() - (timeRange.start_time() - t);
                    }
                    else if (t > timeRange.end_time_exclusive())
                    {
                        t = timeRange.start_time() + (t - timeRange.end_time_exclusive());
                    }
                    player->seek(t);
                }
                break;
            case Private::MouseMode::Picker:
                if (auto app = p.app.lock())
                {
                    const ftk::Box2I& g = getGeometry();
                    const ftk::V2I pos = event.pos - g.min;
                    if (p.samplePos->setIfChanged(pos))
                    {
                        p.colorSample->setIfChanged(getColorSample(pos));
                        p.pick->setIfChanged((pos - getViewPos()) / getViewZoom());
                        _hudUpdate();
                    }
                }
                break;
            default: break;
            }
        }

        void Viewport::mousePressEvent(ftk::MouseClickEvent& event)
        {
            tl::ui::Viewport::mousePressEvent(event);
            FTK_P();
            if (p.pickBinding.button == event.button &&
                ftk::checkKeyModifier(p.pickBinding.modifier, event.modifiers))
            {
                p.mouse.mode = Private::MouseMode::Picker;
                const ftk::Box2I& g = getGeometry();
                const ftk::V2I pos = event.pos - g.min;
                if (p.samplePos->setIfChanged(pos))
                {
                    p.colorSample->setIfChanged(getColorSample(pos));
                    p.pick->setIfChanged((pos - getViewPos()) / getViewZoom());
                    _hudUpdate();
                }
            }
            else if (p.frameShuttleBinding.button == event.button &&
                ftk::checkKeyModifier(p.frameShuttleBinding.modifier, event.modifiers))
            {
                p.mouse.mode = Private::MouseMode::Shuttle;
                if (auto player = getPlayer())
                {
                    player->stop();
                    p.mouse.shuttleStart = player->getCurrentTime();
                }
            }
        }

        void Viewport::mouseReleaseEvent(ftk::MouseClickEvent& event)
        {
            tl::ui::Viewport::mouseReleaseEvent(event);
            FTK_P();
            p.mouse = Private::MouseData();
        }

        void Viewport::_videoUpdate()
        {
            FTK_P();
            std::vector<ftk::ImageOptions> imageOptions;
            std::vector<tl::DisplayOptions> displayOptions;
            for (size_t i = 0; i < p.videoFramesSize; ++i)
            {
                imageOptions.push_back(p.imageOptions);
                displayOptions.push_back(p.displayOptions);
            }
            setImageOptions(imageOptions);
            setDisplayOptions(displayOptions);
        }

        void Viewport::_hudUpdate()
        {
            FTK_P();

            std::string s = p.path.getFileName();
            p.fileNameLabel->setText(!s.empty() ? s : "(No file)");

            s = std::string();
            if (auto app = p.app.lock())
            {
                auto timeUnitsModel = app->getTimeUnitsModel();
                s = timeUnitsModel->getLabel(p.currentTime);
            }
            p.timeLabel->setText(ftk::Format("Time: {0}, {1} FPS, {2} dropped").
                arg(s).
                arg(p.fps, 2, 5).
                arg(static_cast<int>(p.droppedFrames), 3));

            const auto& colorSample = p.colorSample->get();
            p.colorPickerSwatch->setColor(colorSample);
            p.colorPickerLabel->setText(
                ftk::Format("Color: {0} {1} {2} {3}, Pixel: {4}").
                arg(colorSample.r, 2).
                arg(colorSample.g, 2).
                arg(colorSample.b, 2).
                arg(colorSample.a, 2).
                arg(p.pick->get()));

            p.cacheLabel->setText(
                ftk::Format("Cache: {0}% V, {1}% A").
                arg(static_cast<int>(p.cacheInfo.videoPercentage), 3).
                arg(static_cast<int>(p.cacheInfo.audioPercentage), 3));

            p.hudLayout->setVisible(p.hud);
        }
    }
}

File: examples/python/Viewport.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the DJV project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl
import djvPy as djv

import Util

import weakref

class Widget(ftk.IContainer):
    """
    This widget stacks the HUD (heads up display) over the viewport. The
    C++ application subclasses the viewport for this; the Python
    application wraps it instead. The color picker and render items need
    the C++ viewport's picking, so they are not shown here.
    """
    def __init__(self, context, app, parent = None):
        ftk.IContainer.__init__(self, context, "Viewport.Widget", parent)

        self._appWeak = weakref.ref(app)
        self._player = None
        self._fps = 0.0
        self._droppedFrames = 0
        self._currentTime = None
        self._cacheInfo = None

        self._viewport = tl.ui.Viewport(context)
        self._setWidget(self._viewport)

        self._fileNameLabel = ftk.Label(context)
        self._fileNameLabel.font = ftk.FontType.Mono
        self._fileNameLabel.marginRole = ftk.SizeRole.MarginSmall
        self._timeLabel = ftk.Label(context)
        self._timeLabel.font = ftk.FontType.Mono
        self._timeLabel.marginRole = ftk.SizeRole.MarginSmall
        self._infoLabel = ftk.Label(context)
        self._infoLabel.font = ftk.FontType.Mono
        self._infoLabel.marginRole = ftk.SizeRole.MarginSmall
        self._cacheLabel = ftk.Label(context)
        self._cacheLabel.font = ftk.FontType.Mono
        self._cacheLabel.marginRole = ftk.SizeRole.MarginSmall
        self._viewZoomLabel = ftk.Label(context)
        self._viewZoomLabel.font = ftk.FontType.Mono
        self._viewZoomLabel.marginRole = ftk.SizeRole.MarginSmall
        self._hudWidgets = {
            djv.models.HUDItem.FileName: self._fileNameLabel,
            djv.models.HUDItem.Time: self._timeLabel,
            djv.models.HUDItem.Info: self._infoLabel,
            djv.models.HUDItem.Cache: self._cacheLabel,
            djv.models.HUDItem.ViewZoom: self._viewZoomLabel,
        }

        # One layout per corner, arranged like the C++ viewport's HUD.
        self._hudLayout = ftk.VerticalLayout(context)
        self._hudLayout.parent = self
        self._hudLayout.marginRole = ftk.SizeRole.MarginSmall
        self._hudLayout.spacingRole = ftk.SizeRole._None
        self._corners = {}
        for pos in [djv.models.HUDPos.TopLeft, djv.models.HUDPos.TopRight,
                    djv.models.HUDPos.BottomLeft, djv.models.HUDPos.BottomRight]:
            layout = ftk.VerticalLayout(context)
            layout.marginRole = ftk.SizeRole.MarginInside
            layout.spacingRole = ftk.SizeRole._None
            layout.backgroundColor = ftk.ColorRole.Overlay
            self._corners[pos] = layout
        self._corners[djv.models.HUDPos.TopLeft].vAlign = ftk.VAlign.Top
        self._corners[djv.models.HUDPos.TopRight].vAlign = ftk.VAlign.Top
        self._corners[djv.models.HUDPos.BottomLeft].vAlign = ftk.VAlign.Bottom
        self._corners[djv.models.HUDPos.BottomRight].vAlign = ftk.VAlign.Bottom

        topLayout = ftk.HorizontalLayout(context, self._hudLayout)
        topLayout.spacingRole = ftk.SizeRole.SpacingSmall
        topLayout.vAlign = ftk.VAlign.Top
        self._corners[djv.models.HUDPos.TopLeft].parent = topLayout
        spacer = ftk.Spacer(context, ftk.Orientation.Horizontal, topLayout)
        spacer.hStretch = ftk.Stretch.Expanding
        self._corners[djv.models.HUDPos.TopRight].parent = topLayout

        spacer = ftk.Spacer(context, ftk.Orientation.Vertical, self._hudLayout)
        spacer.vStretch = ftk.Stretch.Expanding

        bottomLayout = ftk.HorizontalLayout(context, self._hudLayout)
        bottomLayout.spacingRole = ftk.SizeRole.SpacingSmall
        bottomLayout.vAlign = ftk.VAlign.Bottom
        self._corners[djv.models.HUDPos.BottomLeft].parent = bottomLayout
        spacer = ftk.Spacer(context, ftk.Orientation.Horizontal, bottomLayout)
        spacer.hStretch = ftk.Stretch.Expanding
        self._corners[djv.models.HUDPos.BottomRight].parent = bottomLayout

        selfWeak = weakref.ref(self)
        self._fpsObserver = ftk.DoubleObserver(
            self._viewport.observeFPS,
            lambda value: selfWeak()._fpsUpdate(value))
        self._droppedFramesObserver = ftk.SizeTObserver(
            self._viewport.observeDroppedFrames,
            lambda value: selfWeak()._droppedFramesUpdate(value))
        self._zoomObserver = ftk.DoubleObserver(
            self._viewport.observeZoom,
            lambda value: selfWeak()._zoomUpdate(value))
        self._hudOptionsObserver = djv.models.HUDOptionsObserver(
            app.getViewportModel().observeHUDOptions,
            lambda value: selfWeak()._hudOptionsUpdate(value))

    def getViewport(self):
        return self._viewport

    def setPlayer(self, player):
        self._player = player
        self._viewport.player = player
        if player:
            selfWeak = weakref.ref(self)
            self._currentTimeObserver = tl.RationalTimeObserver(
                player.observeCurrentTime,
                lambda value: selfWeak()._currentTimeUpdate(value))
            self._cacheInfoObserver = tl.PlayerCacheInfoObserver(
                player.observeCacheInfo,
                lambda value: selfWeak()._cacheInfoUpdate(value))
        else:
            self._currentTime = None
            self._cacheInfo = None
            self._currentTimeObserver = None
            self._cacheInfoObserver = None
        self._hudUpdate()

    def setGeometry(self, value):
        ftk.IContainer.setGeometry(self, value)
        self._hudLayout.setGeometry(value)

    def _fpsUpdate(self, value):
        self._fps = value
        self._hudUpdate()

    def _droppedFramesUpdate(self, value):
        self._droppedFrames = value
        self._hudUpdate()

    def _zoomUpdate(self, value):
        self._viewZoomLabel.text = "Zoom: {:.2f}".format(value)

    def _currentTimeUpdate(self, value):
        self._currentTime = value
        self._hudUpdate()

    def _cacheInfoUpdate(self, value):
        self._cacheInfo = value
        self._hudUpdate()

    def _hudOptionsUpdate(self, options):
        self._hudLayout.setVisible(options.enabled)
        for item, widget in self._hudWidgets.items():
            pos = options.items.get(item, djv.models.HUDPos._None)
            widget.parent = self._corners.get(pos, None)

    def _hudUpdate(self):

        player = self._player
        path = player.path if player else None
        self._fileNameLabel.text = path.fileName if path else "(No file)"

        info = []
        ioInfo = player.ioInfo if player else None
        if ioInfo and ioInfo.video:
            videoInfo = ioInfo.video[0]
            info.append("V: {}x{}:{:.2f} {}".format(
                videoInfo.size.w,
                videoInfo.size.h,
                videoInfo.aspect,
                ftk.to_string(videoInfo.type)))
        if ioInfo and ioInfo.audio.isValid:
            info.append("A: {} {} {}".format(
                ioInfo.audio.channelCount,
                tl.to_string(ioInfo.audio.type),
                ioInfo.audio.sampleRate))
        self._infoLabel.text = ", ".join(info)
        self._infoLabel.setVisible(len(info) > 0)

        s = ""
        if self._currentTime is not None:
            app = self._appWeak()
            s = app.getTimeUnitsModel().getLabel(self._currentTime)
        if ioInfo and ioInfo.video:
            self._timeLabel.text = "Time: {}, {:6.2f} FPS, {:3d} dropped".format(
                s, self._fps, self._droppedFrames)
        else:
            self._timeLabel.text = "Time: {}".format(s)

        cache = []
        if self._cacheInfo:
            if ioInfo and ioInfo.video:
                cache.append("{:3d}% V".format(int(self._cacheInfo.videoPercentage)))
            if ioInfo and ioInfo.audio.isValid:
                cache.append("{:3d}% A".format(int(self._cacheInfo.audioPercentage)))
        s = "Cache: {}".format(", ".join(cache)) if cache else ""
        self._cacheLabel.text = s
        self._cacheLabel.setVisible(len(s) > 0)
